simulator/src/main/java/com/hedera/block/simulator/grpc/impl/PublishStreamGrpcServerImpl.java

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: Apache-2.0
+package com.hedera.block.simulator.grpc.impl;
+
+import static com.hedera.block.simulator.metrics.SimulatorMetricTypes.Counter.LiveBlocksProcessed;
+import static java.lang.System.Logger.Level.ERROR;
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.block.simulator.config.data.BlockStreamConfig;
+import com.hedera.block.simulator.config.data.GrpcConfig;
+import com.hedera.block.simulator.grpc.PublishStreamGrpcServer;
+import com.hedera.block.simulator.metrics.MetricsService;
+import com.hedera.hapi.block.protoc.BlockStreamServiceGrpc;
+import com.hedera.hapi.block.protoc.PublishStreamRequest;
+import com.hedera.hapi.block.protoc.PublishStreamResponse;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import io.grpc.Server;
+import io.grpc.ServerBuilder;
+import io.grpc.stub.StreamObserver;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.List;
+import javax.inject.Inject;
+
+/**
+ * Implementation of {@link PublishStreamGrpcServer} that handles incoming block stream publications
+ * via gRPC streaming. This implementation manages the server setup, handles incoming block streams,
+ * tracks processed blocks, and maintains a history of stream statuses. It provides functionality
+ * to start, monitor, and shutdown the gRPC server.
+ */
+public class PublishStreamGrpcServerImpl implements PublishStreamGrpcServer {
```

Comment:
All implementations could be unified under one interface, that means that we close the idea of starting couple of implementations under one instance. Which is okay, judging by the current and future needs.